regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/RedosFinder.java

```diff
@@ -0,0 +1,324 @@
+      regexContainsBackReference = true;
+    }
+
+    private boolean canFail(AutomatonState state) {
+      return canFail(state, !isUsedForFullMatch && !isAnchoredAtEnd(state));
+    }
+
+    private boolean canFail(AutomatonState state, boolean succeedOnEnd) {
+      if (canFailCache.containsKey(state)) {
+        return canFailCache.get(state);
+      }
+      canFailCache.put(state, true);
+      if (state.incomingTransitionType() != AutomatonState.TransitionType.EPSILON) {
+        return true;
+      }
+      if (canMatchAnything(state)) {
+        succeedOnEnd = true;
+        state = state.continuation();
+      }
+      if ((succeedOnEnd && canReachWithoutConsumingInput(state, endOfRegex))) {
+        canFailCache.put(state, false);
+        return false;
+      }
+      for (AutomatonState successor : state.successors()) {
+        if (!canFail(successor, succeedOnEnd)) {
+          canFailCache.put(state, false);
+          return false;
+        }
+      }
+      return true;
+    }
+
+    private boolean canMatchAnything(AutomatonState state) {
+      if (!(state instanceof RepetitionTree)) {
+        return false;
+      }
+      RepetitionTree repetition = (RepetitionTree) state;
+      return repetition.getQuantifier().getMinimumRepetitions() == 0 && repetition.getQuantifier().isOpenEnded()
+        && canMatchAnyCharacter(repetition.getElement());
+    }
+
+    private boolean canMatchAnyCharacter(RegexTree tree) {
+      SimplifiedRegexCharacterClass characterClass = new SimplifiedRegexCharacterClass();
+      for (RegexTree singleCharacter : collectSingleCharacters(tree, new ArrayList<>())) {
+        if (singleCharacter.is(RegexTree.Kind.DOT)) {
+          characterClass.add((DotTree) singleCharacter);
+        } else {
+          characterClass.add((CharacterClassElementTree) singleCharacter);
+        }
+      }
+      return characterClass.matchesAnyCharacter();
+    }
+
+    private List<RegexTree> collectSingleCharacters(@Nullable RegexTree tree, List<RegexTree> accumulator) {
+      if (tree == null) {
+        return accumulator;
+      }
+      if (tree instanceof CharacterClassElementTree || tree.is(RegexTree.Kind.DOT)) {
+        accumulator.add(tree);
+      } else if (tree.is(RegexTree.Kind.DISJUNCTION)) {
+        for (RegexTree alternative : ((DisjunctionTree) tree).getAlternatives()) {
+          collectSingleCharacters(alternative, accumulator);
+        }
+      } else if (tree instanceof GroupTree) {
+        collectSingleCharacters(((GroupTree) tree).getElement(), accumulator);
+      } else if (tree.is(RegexTree.Kind.REPETITION)) {
+        RepetitionTree repetition = (RepetitionTree) tree;
+        if (repetition.getQuantifier().getMinimumRepetitions() <= 1) {
+          collectSingleCharacters(repetition.getElement(), accumulator);
+        }
+      }
+      return accumulator;
+    }
+
+  }
+
+  private void addBacktracking(BacktrackingType newBacktrackingType) {
```

Comment:
it is easy to forget that ordering of enum elements is affects the logic - I would suggest to create enums field showing the ordering priority like:
```java
enum BacktrackingType implements Comparable<BacktrackingType> {
    ALWAYS_EXPONENTIAL(0),
    QUADRATIC_WHEN_OPTIMIZED(1),
    ALWAYS_QUADRATIC(2),
    LINEAR_WHEN_OPTIMIZED(3),
    NO_ISSUE(4);

    private final int priority;

    BacktrackingType(int priority) {
      this.priority = priority;
    }

    public BacktrackingType max(BacktrackingType another) {
      return this.priority > another.priority ? this : another;
    }

  }

```

and then use it in `addBacktracking` method like:
```java
private void addBacktracking(BacktrackingType newBacktrackingType) {
  foundBacktrackingType = foundBacktrackingType.max(newBacktrackingType);
}
```